### PR TITLE
Scripts: Switch to Prettier 2.5 and change formatting around spacing

### LINKED ISSUE #21872: Scripts: Prettier fork not found when running lint-js
**Describe the bug**

Error with the `prettier` dependency while running the lint script from `@wordpress/scripts` after a new `npm install`.

**To reproduce**
1. Create a project and add the `@wordpress/scripts` dev dependency
```
> mkdir scripts-test \
&& cd scripts-test \
&& npm init -y && npm i -D @wordpress/scripts
```
2. Add the script to the package.json: `"lint:js": "wp-scripts lint-js"`
3. Create a js file and run the lint to see it is working.
```
> echo 'const x = 1;' >index.js \
&& npm run lint:js
```
4. Checking the prettier dependency in this moment:
```
> npm ls prettier
scripts-test@1.0.0 /Users/renatho/Downloads/scripts-test
└─┬ @wordpress/scripts@7.1.3
  ├─┬ @wordpress/eslint-plugin@4.0.0
  │ └── prettier@npm:wp-prettier@1.19.1  deduped
  └── prettier@npm:wp-prettier@1.19.1
```
5. You can make a `git init` here and commit the `package-lock` to see the next changes.
6. Remove the `node_modules` folder, reinstall the packages and run the lint script. The new install will update (unexpectedly) our `package-lock.json` and It stops to work because it doesn't find the `prettier`. You will see the error `Error: Cannot find module 'prettier'`
```
> rm -rf node_modules \
&& npm i \
&& npm run lint:js
```
7. Check the prettier dependency to understand what happened (you can also see the `package-lock` diff):
```
> npm ls prettier
scripts-test@1.0.0 /Users/renatho/Downloads/scripts-test
└─┬ @wordpress/scripts@7.1.3
  ├─┬ @wordpress/eslint-plugin@4.0.0
  │ └── UNMET PEER DEPENDENCY prettier@npm:wp-prettier@1.19.1 
  └── prettier@npm:wp-prettier@1.19.1
```

**Expected behavior**

Install the `@wordpress/scripts` and after other `npm install` it continues to work properly while running the `npm run lint:js`.

**Version (please complete the following information):**
- A clean project with only the `@wordpress/scripts` installed. Reproduced in `7.1.3` and `8.0.1`

**Desktop (please complete the following information):**
 - OS: macOS Catalina
 - NPM 6.14.4 / Node 10.17.0

### LINKED ISSUE #37516: @wordpress/prettier-config 's jsxBracketSameLine config is now deprecated
### Description

When setting up our project's esllint using @wordpress/eslint-plugin, specified in `.eslintrc` as per:
`"extends": [ "plugin:@wordpress/eslint-plugin/recommended" ],`

I get this error notice `jsxBracketSameLine is deprecated.`

I found that this error notice is Prettier related, and comes from [this line in @wordpress/prettier-config](https://github.com/WordPress/gutenberg/blob/037fcc79749b186e01ba34f3104971aebc3288dc/packages/prettier-config/lib/index.js#L21)

This change is from Prettier 2.4
https://prettier.io/blog/2021/09/09/2.4.0.html

> This release renames the jsxBracketSameLine option to bracketSameLine, which supports HTML, Vue, and Angular in addition to JSX. The old name has been deprecated.

@wordpress/prettier-config's config setting needs to be updated

### Step-by-step reproduction instructions

Here, we are using @wordpress/eslint-plugin as standalone -- not as part of @wordpress/scripts

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by gziolo**
## Description

**TLDR; I propose we switch to the original version of Prettier and update it to the most recent version 2.5. Consequently, we should also update the existing JavaScript Coding Standards around spacing to meet what Prettier can do.**

Fixes #21872.
Fixes #37516.
Replaces and closes #37517.

We use an outdated fork of Prettier in the Gutenberg project because it didn't get any updates for more than a year:

| `wp-prettier`  | `prettier` |
| ------------- | ------------- |
| Screenshot 2021-12-23 at 09 40 31  | Screenshot 2021-12-23 at 09 42 18 |

The only motivation behind using a forked version of Prettier is its custom functionality overriding formatting behavior for spaces. It allows to align closer with the coding style convention used for PHP in the WordPress codebase as explained in [JavaScript Coding Standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#spacing). Related rules:

> Use spaces liberally throughout your code. “When in doubt, space it out.”

> The WordPress JavaScript standards prefer slightly broader whitespace rules than the jQuery style guide. These deviations are for consistency between the PHP and JavaScript files in the WordPress codebase.

It's worth reminding that this topic was discussed several times, beginning in 2017 when the initial exploration of using Prettier started with #2819. Throughout all that time, the Prettier maintainers didn't change their mind about adding support for spaces between parens (`()`) and brackets (`[]`) despite many requests shared in https://github.com/prettier/prettier/issues/1303 from users (including the WordPress community). That's why we decided to use the forked version developed by developers at Automattic.

In the meantime, there were three minor releases of Prettier with a long list of new features in the last year:

- [Prettier 2.5: TypeScript 4.5 and MDX v2 comment syntax!](https://prettier.io/blog/2021/11/25/2.5.0.html)
- [Prettier 2.4: new bracketSameLine option and TypeScript 4.4 support!](https://prettier.io/blog/2021/09/09/2.4.0.html)
- [Prettier 2.3. In which assignments are consistent, short keys non-breaking, and Handlebars official](https://prettier.io/blog/2021/05/09/2.3.0.html)

The Gutenberg project no longer uses only JavaScript for development. Some parts of the codebase use TypeScript, which [we agreed upon a few months back](https://make.wordpress.org/core/2021/08/11/finalizing-the-native-typescript-proposal/). It means we can't use improvements added for newer versions of the TypeScript language and potentially for future additions to the JavaScript language. A deprecation was also added to one of the config options we use in the shared Prettier config that the community uses. It turns out that it prevents using `@wordpress/prettier-config` with the most recent version of Prettier in 3rd party projects, as explained in #37516. Finally, we had multiple reports in #21872 from the community members using `@wordpress/scripts` or `@wordpress/eslint-plugin` that the forked version of Prettier doesn't install correctly in their projects. The good news is that switching to the original Prettier will resolve all of the above!

## JavaScript Coding Standard changes necessary

```diff
- Some whitespace rules differ, for consistency with the WordPress PHP coding standards.
+ Some whitespace rules differ, for consistency with the [Prettier](https://prettier.io) formatting tool.
```

```diff
- All new or updated JavaScript code will be reviewed to ensure it conforms to the standards, and passes JSHint.
+ All new or updated JavaScript code will be reviewed to ensure it conforms to the standards, and passes [ESLint](https://eslint.org).
```

```diff
- Use spaces liberally throughout your code. “When in doubt, space it out.”
```

```diff
- Any ! negation operator should have a following space.*
+ Any ! negation operator should not have a following space.
```

```diff
- Always include extra spaces around elements and arguments:
- array = [ a, b ]; 
- foo( arg );
- foo( 'string', object );
- foo( options, object[ property ] );
- foo( node, 'property', 2 );
- prop = object[ 'default' ];
- firstArrayElement = arr[ 0 ];
+ Never include extra spaces around elements and arguments: 
+ arg = [a, b]; 
+ foo(arg);
+ foo('string', object);
+ foo(options, object[property]);
+ foo(node, 'property', 2);
+ prop = object['default'];
+ firstArrayElement = arr[0];
```

```diff
- var i;
-
- if ( condition ) {
-    doSomething( 'with a string' );
- } else if ( otherCondition ) {
-     otherThing( {
-         key: value,
-         otherKey: otherValue
-     } );
- } else {
-     somethingElse( true );
- }
 
- // Unlike jQuery, WordPress prefers a space after the ! negation operator.
- // This is also done to conform to our PHP standards.
- while ( ! condition ) {
-     iterating++;
- }
-  
- for ( i = 0; i < 100; i++ ) {
-     object[ array[ i ] ] = someFn( i );
-     $( '.container' ).val( array[ i ] );
- }
-  
- try {
-     // Expressions
- } catch ( e ) {
-     // Expressions
- }
+ var i;
+
+ if (condition) {
+ 	doSomething('with a string');
+ } else if (otherCondition) {
+ 	otherThing({
+ 		key: value,
+ 		otherKey: otherValue,
+ 	});
+ } else {
+ 	somethingElse(true);
+ }
+ 
+ while (!condition) {
+ 	iterating++;
+ }
+ 
+ for (i = 0; i < 100; i++) {
+ 	object[array[i]] = someFn(i);
+ 	$('.container').val(array[i]);
+ }
+ 
+ try {
+ 	// Expressions
+ } catch (e) {
+ 	// Expressions
+ }
```

There are more code examples that we would have to update in the document, but they are very similar to those shared above.

## How has this been tested?

```bash
npm install
npm run build
npm run lint-js
npm run test-unit
npm run test-e2e
```

I also ensured that there were no changes in the branch when running:

```bash
npm run docs:build
npm run format
```

## Types of changes

Breaking change (fix or feature that would cause existing functionality not to work as expected).

## TODO tasks

- [x] Bring this topic again to the next Core JS office hours.
- [x] Post an official proposal at https://make.wordpress.org/core
- [ ] Prepare all the required changes to the [JavaScript Coding Standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/) for WordPress.
- [ ] Add CHANGELOG entries for breaking changes that impact WordPress packages.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [-] My code follows the accessibility standards. 
- [-] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by Mamaduka** [comment]
I'm also worried about how similar changes affect git ~~blame~~ history. I use GitLens on VS Code daily to find context around specific decisions or PRs that introduced them. Unfortunately, large-scale changes like this make it a little harder.

Yesterday, I came across this conversation on [Twitter](https://twitter.com/acemarke/status/1478060904562536450). So maybe we should also consider adding `.git-blame-ignore-revs` to a project.

Another commit we can consider for that list is 4857ad58c1241b3d63d21a6880c989b85746c3dc.

P.S. Article linked in a tweet - https://akrabat.com/ignoring-revisions-with-git-blame/

**Comment by gziolo** [comment]
I have just posted an official proposal at https://make.wordpress.org/core/2022/01/05/proposal-changes-to-javascript-coding-standards-for-full-prettier-compatibility/.

**Comment by kevin940726** [comment]
I'm all in for aligning to the community's standard. However, wouldn't it be more backward-compatible if we continue to follow the WordPress style guide but using `eslint --fix` in conjunction with `prettier --write`?

So instead of running our own fork of prettier with `wp-prettier --write`, we can do `prettier --write && eslint --fix` to fix all those minor issues that prettier doesn't care about.

Sure, it's gonna be slower, but I don't think we'd be running that command that often, and it can be automated with `lint-staged` as well. IDE can also be configured to run both tools in order (or just disable them linting styling rules in IDE, it shouldn't care afterall).

I agree that just changing the rules once and for all is a much straightforward solution though, but I also have the concern brought up by @Mamaduka in https://github.com/WordPress/gutenberg/pull/37607#issuecomment-1004558433 that git blaming would be a pain to deal with.

**Comment by gziolo** [comment]
> However, wouldn't it be more backward-compatible if we continue to follow the WordPress style guide but using eslint --fix in conjunction with prettier --write?

@ntwb did several explorations in https://github.com/WordPress/gutenberg/pull/4628 of the same approach you mentioned, but it wasn't good enough back then in 2018.

**Comment by kevin940726** [comment]
> @ntwb did several explorations in #4628 of the same approach you mentioned, but it wasn't good enough back then in 2018.

Is the reason still relevant today? Seems like the conclusion in #4628 is to not use prettier which has obviously changed now. Might be worth noting that I've tried using `prettier --write && eslint --fix` together before in one of my projects and it works really well.

**Comment by ntwb** [comment]
> Is the reason still relevant today? Seems like the conclusion in #4628 is to not use prettier which has obviously changed now. Might be worth noting that I've tried using prettier --write && eslint --fix together before in one of my projects and it works really well.

I think this would be an interesting task to explore, wondering if we can create a dedicated ESLint configuration that only fixes "spaces between parens and brackets", and if running lint-staging with Prettier first and then this ESLint config would work....

It seems pretty hacky to me to be honest, and not very friendly to the wider developer community to have to implement it in this manner,not sure if `@wordpress/scripts` could handle this either....

**Comment by gziolo** [comment]
> It seems pretty hacky to me to be honest, and not very friendly to the wider developer community to have to implement it in this manner,not sure if @wordpress/scripts could handle this either....

As long people would use `npx wp-scripts format` it would be fine, but in case someone would use Prettier directly they would get different styles, for example when using the default IDE integration.

**Comment by kevin940726** [comment]
> As long people would use `npx wp-scripts format` it would be fine, but in case someone would use Prettier directly they would get different styles, for example when using the default IDE integration.

True, but that shouldn't matter either though. People should be able to use any styling during development, as long as the code they pushed is following WP standard, that's the reason we have `lint-staged` in the first place.

> wondering if we can create a dedicated ESLint configuration that only fixes "spaces between parens and brackets"

In theory, it should be possible, but I don't think we necessarily have to do this, it only has some performance benefits.

> and if running lint-staging with Prettier first and then this ESLint config would work....

[It should be possible.](https://github.com/okonet/lint-staged#running-multiple-commands-in-a-sequence)

**Comment by luisherranz** [comment]
`lint-staged` is a great tool, but I think IDE integrations should not be underestimated. For many developers (me included) it's an essential workflow:

- Write some code.
- Use IDE shortcut to run autoformat (for example, Prettier integration).
- Write more code.
- Use IDE shortcut to run autoformat again.
- Repeat...

If the correct formating is only achieved using a CLI command like `npx wp-scripts format`, the developer experience won't be that good.

**Comment by kevin940726** [comment]
> If the correct formating is only achieved using a CLI command like `npx wp-scripts format`, the developer experience won't be that good.

My imagination would be that `eslint` will only be fixing some minor details (like spaces around parenthesis and such), so the overall styling is still pretty much the same. In addition, we'll still be providing prettier config and eslint config, so the existing IDE integration should still work. By any means, not enforcing formatting styling during development should be better for DX, not worse, so that developers can use whatever styling they feel comfortable with, as long as the final result is following the standard.

Moreover, it should be possible to configure IDE integration to run `prettier --write && eslint --fix` too, so that CLI isn't the only achievable method. But maybe each IDE has some quirks though 😅 .

**Comment by luisherranz** [comment]
> it should be possible to configure IDE integration to run prettier --write && eslint --fix too, so that CLI isn't the only achievable method

Yeah, I guess it is possible. But each step we deviate from the standard workflow is usually an additional step users need to do to configure the tooling. In that case, they would need to figure out how to run that command instead of just installing a Prettier plugin.

**Comment by ntwb** [comment]
Thanks, I think it's worth exploring, if the workflows can be solved, without large amounts of pain, so that the Gutenberg Repo (Commit, Lint-Staged, GitHub Actions etc) and Developer IDE/Editor work (Prettier/ESLint save & format) 

If the above can be achieved, then Gutenberg no longer needs to use (or update & maintain) the Prettier fork and can use the canonical Prettier then that covers I think all of the main objections to the make/core proposal here:

- https://make.wordpress.org/core/2022/01/05/proposal-changes-to-javascript-coding-standards-for-full-prettier-compatibility/

**Comment by kevin940726** [comment]
I did some quick testing, but it turns out there are unfortunately some rules that won't be able to stay intact. There's the [`space-in-parens`](https://eslint.org/docs/2.0.0/rules/space-in-parens) rule that will fix the prettier spaces issue for us, however, that will conflict with the [`printWidth`](https://prettier.io/docs/en/options.html#print-width) rule in prettier. Since that `eslint` will be running after `prettier`, and there's no one-to-one equivalent of `printWidth` in `eslint`, it won't be able to _recover_ from the width change. Take below as an example:

```js
// Original code that's really long and exceeds the `printWidth` setting
fn(
  ...
);

// Prettier: the width is just below the threshold after removing the spaces
fn(...);

// eslint adds back the spaces, but it won't wrap them into multiple lines anymore
fn( 123 );
```

This creates some inconsistencies between the modified code and the original code. Even though the impact would be much lower than what this PR proposes, we still have to solve the same problem, just on a smaller scale.

> Yeah, I guess it is possible. But each step we deviate from the standard workflow is usually an additional step users need to do to configure the tooling. In that case, they would need to figure out how to run that command instead of just installing a Prettier plugin.

We can provide documentation just like how we do [currently](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/getting-started-with-code-contribution.md#developer-tools). They don't "need" to though, I think it's an important point, the whole IDE setup should be optional.

**Comment by gziolo** [comment]
I also found PR where `prettier-eslint` was explored: https://github.com/WordPress/gutenberg/pull/4374.

> I did some quick testing, but it turns out there are unfortunately some rules that won't be able to stay intact. There's the [`space-in-parens`](https://eslint.org/docs/2.0.0/rules/space-in-parens) rule that will fix the prettier spaces issue for us, however, that will conflict with the [`printWidth`](https://prettier.io/docs/en/options.html#print-width) rule in prettier. 

Yes, that was one of the concerns discussed during the initial explorations of Prettier. ESLint operates after Prettier, and it doesn't respect the line width setting. The same issue likely applies to the `wp-prettier`'s fork. 

Looking at the `prettier-eslint` solution more broadly, I think it suffers from the same root issue as `wp-prettier` – it’s a different external thing that needs to be maintained. In this case, we can also suffer from the lack of updates to the project. I checked the latest release [13.0.0](https://unpkg.com/browse/prettier-eslint@13.0.0/package.json) (6 months ago), and I discovered it depends on ESLint `^7.9.0`, and we already use the next major version (`8.6.0`), so we would suddenly ship two different major versions of ESLint with `@wordpress/scripts`. I’m sure you could limit `--fix` to rules impacting spacing as mentioned in one of the comments, so it would most likely work fine after all. The fact that we use TypeScript more and more in the project brings another set of challenges because we would have to use fixers from TS rules to cover the entire codebase rather than those that ESLint ships by default.

Anyway, you usually modify only a few JS files so that you won’t notice any issues. However, we use also Prettier in Markdown files. It’s getting more interesting there because Prettier is smart enough to detect JS snippets there and format them. However, ESLint knows nothing about Markdown, so we would diverge there, or we would have to ignore those snippets when using Prettier, so they don't get formatted without the ESLint step that fixes spaces.

The most crucial part is optimizing the developer experience because many devs use IDE integrations. I know that the best part is that you just put a Prettier's config and everything magically works in Visual Studio Code when you save your code, and I suspect it's not that different experience in other editors listed at https://prettier.io/docs/en/editors.html. 

While Prettier + ESLint seems like a viable solution, I don't think it is clearly better than the current setup with the `wp-pretttier` fork. They both have many disadvantages and keep us in a place where we need to workaround the limitations of Prettier. I personally would prefer to either change the coding styles as proposed or leave everything as is and eventually try to update the fork.

**Comment by simonhammes** [comment]
@gziolo Sorry for bothering you, but are there any news on the proposal/this PR?

**Comment by gziolo** [comment]
I'm re-sharing the comment from @ntwb that he left on WordPress Slack (link requires registration at https://make.wordpress.org/chat):

https://wordpress.slack.com/archives/C5UNMSU4R/p1648555187344279?thread_ts=1647357562.940759&cid=C5UNMSU4R

> The consensus at the moment is a “No” from the core team.
> I’ve a draft comment at the moment that I’ll post tomorrow, but the tl;dr of this is I think the WordPress project should resubmit it’s request upstream to Prettier in a more formal manner, and I’ll add why I think we might be able to achieve a better result this time

I'm closing this PR becaue of that.

**Comment by gziolo** [comment]
I opened a new issue in the Prettier repository that proposed again the option of allowing extra spaces around elements and arguments in JavaScript code: https://github.com/prettier/prettier/issues/13107.
